compiler/src/dotty/tools/dotc/cc/CaptureOps.scala

```diff
@@ -0,0 +1,168 @@
+
+extension (tree: Tree)
+
+  /** Map tree with CaptureRef type to its type, throw IllegalCaptureRef otherwise */
+  def toCaptureRef(using Context): CaptureRef = tree.tpe match
+    case ref: CaptureRef => ref
+    case tpe => throw IllegalCaptureRef(tpe)
+
+  /** Convert a @retains or @retainsByName annotation tree to the capture set it represents.
+   *  For efficience, the result is cached as an Attachment on the tree.
+   */
+  def toCaptureSet(using Context): CaptureSet =
+    tree.getAttachment(Captures) match
+      case Some(refs) => refs
+      case None =>
+        val refs = CaptureSet(retainedElems(tree).map(_.toCaptureRef)*)
+          .showing(i"toCaptureSet $tree --> $result", capt)
+        tree.putAttachment(Captures, refs)
+        refs
+
+extension (tp: Type)
+
+  /** @pre `tp` is a CapturingType */
+  def derivedCapturingType(parent: Type, refs: CaptureSet)(using Context): Type = tp match
+    case tp @ CapturingType(p, r) =>
+      if (parent eq p) && (refs eq r) then tp
+      else CapturingType(parent, refs, tp.isBoxed)
+
+  /** If this is a unboxed capturing type with nonempty capture set, its boxed version.
+   *  Or, if type is a TypeBounds of capturing types, the version where the bounds are boxed.
+   *  The identity for all other types.
+   */
+  def boxed(using Context): Type = tp.dealias match
+    case tp @ CapturingType(parent, refs) if !tp.isBoxed && !refs.isAlwaysEmpty =>
+      tp.annot match
+        case ann: CaptureAnnotation =>
+          ann.boxedType(tp)
+        case ann =>
+          ann.tree.getAttachment(BoxedType) match
+            case None => ann.tree.putAttachment(BoxedType, BoxedTypeCache())
+            case _ =>
+          ann.tree.attachment(BoxedType)(tp)
+    case tp: RealTypeBounds =>
+      tp.derivedTypeBounds(tp.lo.boxed, tp.hi.boxed)
+    case _ =>
+      tp
+
+  /** If `sym` is a type parameter, the boxed version of `tp`, otherwise `tp` */
+  def boxedIfTypeParam(sym: Symbol)(using Context) =
+    if sym.is(TypeParam) then tp.boxed else tp
+
+  /** The boxed version of `tp`, unless `tycon` is a function symbol */
+  def boxedUnlessFun(tycon: Type)(using Context) =
+    if ctx.phase != Phases.checkCapturesPhase || defn.isFunctionSymbol(tycon.typeSymbol)
+    then tp
+    else tp.boxed
+
+  /** The capture set consisting of all top-level captures of `tp` that appear under a box.
+   *  Unlike for `boxed` this also considers parents of capture types, unions and
+   *  intersections, and type proxies other than abstract types.
+   */
+  def boxedCaptureSet(using Context): CaptureSet =
+    def getBoxed(tp: Type): CaptureSet = tp match
+      case tp @ CapturingType(parent, refs) =>
+        val pcs = getBoxed(parent)
+        if tp.isBoxed then refs ++ pcs else pcs
+      case tp: TypeRef if tp.symbol.isAbstractType => CaptureSet.empty
+      case tp: TypeProxy => getBoxed(tp.superType)
+      case tp: AndType => getBoxed(tp.tp1) ++ getBoxed(tp.tp2)
```

Comment:
I am not sure whether this is correct. Should we use `++` for unions and `**` for intersections?